lucille-core/src/main/java/com/kmwllc/lucille/stage/CopyFields.java

```diff
@@ -1,70 +1,93 @@
  *   <li>update_mode (String, Optional) : Determines how writing will be handling if the destination field is already populated. Can be
- *   'overwrite', 'append' or 'skip'. Defaults to 'overwrite'.</li>
+ *   'overwrite', 'append' or 'skip'. Defaults to 'overwrite'. Cannot be set with isNested.</li>
+ *   <li>isNested (Boolean, Optional) : Sets whether or not to treat the given field name as nested json or a literal field name.
+ *   It true, the field name will be split on '.' and each part will be treated as a level of nesting. Cannot be set with update_mode.
+ *   Defaults to false. </li>
  * </ul>
  */
 public class CopyFields extends Stage {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  private static final ObjectMapper mapper = new ObjectMapper();
 
   public static final Spec SPEC = SpecBuilder.stage()
-      .requiredList("source", new TypeReference<List<String>>(){})
-      .requiredList("dest", new TypeReference<List<String>>(){})
-      .optionalString("update_mode").build();
+      .requiredParent("fieldMapping", new TypeReference<Map<String, String>>() {})
+      .optionalString("update_mode")
+      .optionalBoolean("isNested").build();
 
-  private final List<String> sourceFields;
-  private final List<String> destFields;
+  private final Map<String, Object> fieldMapping;
   private final UpdateMode updateMode;
+  private final boolean isNested;
 
   public CopyFields(Config config) {
     super(config);
-    this.sourceFields = config.getStringList("source");
-    this.destFields = config.getStringList("dest");
+    this.fieldMapping = config.getConfig("fieldMapping").root().unwrapped();
     this.updateMode = UpdateMode.fromConfig(config);
+    this.isNested = ConfigUtils.getOrDefault(config, "isNested", false);
   }
 
   @Override
   public void start() throws StageException {
-    StageUtils.validateFieldNumNotZero(sourceFields, "Copy Fields");
-    StageUtils.validateFieldNumNotZero(destFields, "Copy Fields");
-    StageUtils.validateFieldNumsSeveralToOne(sourceFields, destFields, "Copy Fields");
+    if (this.fieldMapping.size() == 0) {
+      throw new StageException("fieldMapping must have at least one source-dest pair for Copy Fields");
+    }
+
+    if (this.isNested && this.updateMode != null) {
+      log.info("Cannot set both isNested and update_mode in Copy Fields at the same time. Ignoring update_mode, fields will be overwritten.");
+    }
   }
 
   @Override
   public Iterator<Document> processDocument(Document doc) throws StageException {
-    for (int i = 0; i < sourceFields.size(); i++) {
-      // If there is only one source or dest, use it. Otherwise, use the current source/dest.
-      String sourceField = sourceFields.get(i);
-      String destField = destFields.size() == 1 ? destFields.get(0) : destFields.get(i);
+    for (Entry<String, Object> fieldPair : this.fieldMapping.entrySet()) {
```

Comment:
an optimization for the isNested case would be to convert the source and dest fields to String[] in the start() method and use those String[]'s when calling doc.getNestedJson() and doc.setNestedJson()

this would avoid having to repeatedly String.split the same source and dest fields in getNestedJson/setNestedJson for every single doc